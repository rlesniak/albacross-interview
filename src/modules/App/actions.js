/* @flow */

import api from 'Utils/api';

import type { List } from 'immutable';
import type { ExtractReturn } from './types';

export const SET_FETCHING = 'SET_FETCHING';
export const UNSET_FETCHING = 'UNSET_FETCHING';
export const SET_CURRENCIES = 'SET_CURRENCIES';
export const SET_EXCHANGES = 'SET_EXCHANGES';
export const SET_ACTIVE = 'SET_ACTIVE';
export const ADD_TO_FAV = 'ADD_TO_FAV';

export const setFetching = (name: string) => ({
  type: SET_FETCHING,
  name,
});

export const unsetFetching = (name: string) => ({
  type: UNSET_FETCHING,
  name,
});

export const setCurrencies = (currencies: []) => ({
  type: SET_CURRENCIES,
  currencies,
});

export const setExchanges = (exchanges: []) => ({
  type: SET_EXCHANGES,
  exchanges,
});

export const setAciveCurrency = (currency: string) => ({
  type: SET_ACTIVE,
  currency,
});

export const addToFav = (name: string) => ({
  type: ADD_TO_FAV,
  name,
});

export const fetchCurrencies = () => (dispatch: Dispatch) => {
  const FETCH_NAME = 'currencies';
  dispatch(setFetching(FETCH_NAME));

  api.get('list').then(({ currencies }) => {
    dispatch(setCurrencies(currencies));
    dispatch(unsetFetching(FETCH_NAME));
  });
};

export const fetchExchanges = (currencies: List<string>) => (dispatch: Dispatch) => {
  const FETCH_NAME = 'exchanges';
  dispatch(setFetching(FETCH_NAME));

  api.get('live', `currencies=${currencies.join(',')}`).then(({ quotes, source, success }) => {
    if (success) {
      dispatch(setExchanges(quotes));
      dispatch(setAciveCurrency(source));
    }

    dispatch(unsetFetching(FETCH_NAME));
  });
};

export type ActionsType =
  | ExtractReturn<typeof setFetching>
  | ExtractReturn<typeof unsetFetching>
  | ExtractReturn<typeof setCurrencies>
  | ExtractReturn<typeof setAciveCurrency>
  | ExtractReturn<typeof setExchanges>
  | ExtractReturn<typeof addToFav>;
