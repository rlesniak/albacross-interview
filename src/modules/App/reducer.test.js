import { Map } from 'immutable';

import AppReducer, { INITIAL_STATE } from './reducer';

import * as ACTIONS from './actions';

describe('App reducer', () => {
  it('should return proper initial state', () => {
    const state = AppReducer(undefined, {});

    expect(state).toMatchSnapshot();
  });

  it('should create proper SET_FETCHING action', () => {
    const state = AppReducer(undefined, {
      type: ACTIONS.SET_FETCHING,
      name: 'currencies',
    });

    expect(state).toMatchSnapshot();
  });

  it('should create proper UNSET_FETCHING action', () => {
    const state = AppReducer(
      {
        ...INITIAL_STATE,
        dataFetching: Map({ currencies: true }),
      },
      {
        type: ACTIONS.UNSET_FETCHING,
        name: 'currencies',
      },
    );

    expect(state).toMatchSnapshot();
  });

  it('should create proper SET_CURRENCIES action', () => {
    const state = AppReducer(undefined, {
      type: ACTIONS.SET_CURRENCIES,
      currencies: {
        BYC: 'test1', BTC: 'test2', AED: 'test', PLN: 'test4', USD: 'test3',
      },
    });

    expect(state).toMatchSnapshot();
  });

  it('should create proper SET_EXCHANGES action', () => {
    const state = AppReducer(undefined, {
      type: ACTIONS.SET_EXCHANGES,
      exchanges: {
        BYC: 1.23, BTC: 0.32, PLN: 0.00,
      },
    });

    expect(state).toMatchSnapshot();
  });
});
