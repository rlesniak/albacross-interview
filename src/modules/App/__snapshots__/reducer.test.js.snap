// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App reducer should create proper SET_CURRENCIES action 1`] = `
Object {
  "activeCurrency": "USD",
  "currenciesList": Immutable.OrderedMap {
    "AED": "test",
    "BYC": "test1",
    "BTC": "test2",
    "USD": "test3",
    "PLN": "test4",
  },
  "dataFetching": Immutable.Map {},
  "exchangesList": Immutable.Map {},
  "favoriteCurrencies": Immutable.List [],
}
`;

exports[`App reducer should create proper SET_EXCHANGES action 1`] = `
Object {
  "activeCurrency": "USD",
  "currenciesList": Immutable.Map {},
  "dataFetching": Immutable.Map {},
  "exchangesList": Immutable.OrderedMap {
    "BYC": 1.23,
    "BTC": 0.32,
    "PLN": 0,
  },
  "favoriteCurrencies": Immutable.List [],
}
`;

exports[`App reducer should create proper SET_FETCHING action 1`] = `
Object {
  "activeCurrency": "USD",
  "currenciesList": Immutable.Map {},
  "dataFetching": Immutable.Map {
    "currencies": true,
  },
  "exchangesList": Immutable.Map {},
  "favoriteCurrencies": Immutable.List [],
}
`;

exports[`App reducer should create proper UNSET_FETCHING action 1`] = `
Object {
  "activeCurrency": "USD",
  "currenciesList": Immutable.Map {},
  "dataFetching": Immutable.Map {
    "currencies": false,
  },
  "exchangesList": Immutable.Map {},
  "favoriteCurrencies": Immutable.List [],
}
`;

exports[`App reducer should return proper initial state 1`] = `
Object {
  "activeCurrency": "USD",
  "currenciesList": Immutable.Map {},
  "dataFetching": Immutable.Map {},
  "exchangesList": Immutable.Map {},
  "favoriteCurrencies": Immutable.List [],
}
`;
